### PR TITLE
CPU manager interfaces.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
Blocker for CPU manager #49186 (3 of 6)

* Previous PR in this series: #51132
* Next PR in this series: #51357

@sjenning @derekwaynecarr

### CODE REVIEW

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	// UnregisterContainer is called near UnregisterContainer() so that the
+	// manager stops trying to update that container in the reconcilation loop
+	// and any CPUs dedicated to the container are freed to the shared pool.
+	UnregisterContainer(containerID string) error
+
+	// Returns a read-only interface to the CPU manager internal state.
+	State() state.Reader
+
+	// Returns `true` if the node is under CPU pressure.
+	IsUnderCPUPressure() bool
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: note: this will be removed.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	lifecycle.PodAdmitHandler
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: note: this will be removed.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+type runtimeService interface {
+	UpdateContainerResources(id string, resources *runtimeapi.LinuxContainerResources) error
+}
+
+type policyName string
+
+// Manager interface provides methods for Kubelet to manage pod cpus.
+type Manager interface {
+	// Start is called during Kubelet initialization.
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: can you make clear what prereqs should be satisfied before calling start?  i.e. is cAdvisor up or not?

File: pkg/kubelet/cm/cpumanager/state/state.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+)
+
+// Reader interface used to read current cpu/pod assignment state
+type Reader interface {
+	GetCPUSet(containerID string) (cpuset.CPUSet, bool)
+	GetDefaultCPUSet() cpuset.CPUSet
+	GetCPUSetOrDefault(containerID string) cpuset.CPUSet
+	GetPressure() bool
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/state/state.go]: note: drop this

File: pkg/kubelet/cm/cpumanager/state/state.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+type writer interface {
+	SetCPUSet(containerID string, cpuset cpuset.CPUSet)
+	SetDefaultCPUSet(cpuset cpuset.CPUSet)
+	Delete(containerID string)
+	SetPressure(value bool)
```
Comment (derekwaynecarr) [inline pkg/kubelet/cm/cpumanager/state/state.go]: note: drop this

**Comment by derekwaynecarr** [comment]
remove the pressure bits, then this is lgtm.  poke when done.  thanks

**Comment by ConnorDoyle** [comment]
@derekwaynecarr updated

File: pkg/kubelet/cm/cpumanager/cpu_manager.go (hunk 3 above)
Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: Added precondition

Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go]: Removed precondition again! With the refactored kubelet wiring we are getting machine info passed into the constructor.

File: pkg/kubelet/cm/cpumanager/cpu_manager.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	Start(activePods ActivePodsFunc, podStatusProvider status.PodStatusProvider)
+
+	// AddContainer is called between container create and container start
+	// so that initial CPU affinity settings can be written through to the
+	// container runtime before the first process begins to execute.
+	AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go]: @derekwaynecarr -- as discussed in the next PR for this feature we considered changing the name of these functions from `RegisterContainer` and `UnregisterContainer` to indicate when in the container lifecycle they should be called.

Our current plan is to have an interface for internal container lifecycle callbacks within the container manager. Those functions will have names like `PreStartContainer`, `PreStopContainer`, `PostStopContainer` etc. However, we want `UnregisterContainer` to get called during both `removeContainer` and `killContainer`, so neither `PreStopContainer` nor `PostStopContainer` makes sense...

What we would like to do is call the lifecycle callback `PreStopContainer` from `killContainer` and call `PostStopContainer` from `removeContainer`. In the hooks impl we get from the container manager, both `PreStopContainer` and `PostStopContainer` will call `cpuManager.UnregisterContainer`.

As a compromise, I have renamed `RegisterContainer` => `AddContainer` and `UnregisterContainer` => `RemoveContainer`, and clarified the banner comments. Hope this works for you.

**Comment by ConnorDoyle** [comment]
@derekwaynecarr updated, with a comment about naming

**Comment by derekwaynecarr** [comment]
/approve
/lgtm

thanks for taking the time to work through the call path in https://github.com/kubernetes/kubernetes/pull/51357

this is almost done!

**Comment by ConnorDoyle** [comment]
@derekwaynecarr please add v1.8 milestone

**Pull request merged**